turbo/txpool-provider/control.go

```diff
@@ -0,0 +1,168 @@
+package txpoolprovider
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/ledgerwatch/turbo-geth/common"
+	"github.com/ledgerwatch/turbo-geth/core/rawdb"
+	"github.com/ledgerwatch/turbo-geth/core/types"
+	"github.com/ledgerwatch/turbo-geth/core/types/accounts"
+	"github.com/ledgerwatch/turbo-geth/ethdb"
+	"github.com/ledgerwatch/turbo-geth/ethdb/remote/remotedbserver"
+	"github.com/ledgerwatch/turbo-geth/params"
+	"github.com/ledgerwatch/turbo-geth/turbo/adapter"
+	"github.com/ledgerwatch/turbo-geth/turbo/txpool-provider/pb"
+)
+
+type TxPoolControlServer struct {
+	pb.UnimplementedTxpoolControlServer
+	chainConfig *params.ChainConfig
+	kv          ethdb.KV
+	events      *remotedbserver.Events
+}
+
+func NewTxPoolControlServer(cc *params.ChainConfig, kv ethdb.KV, events *remotedbserver.Events) *TxPoolControlServer {
+	return &TxPoolControlServer{chainConfig: cc, kv: kv, events: events}
+}
+
+func (c *TxPoolControlServer) AccountInfo(ctx context.Context, request *pb.AccountInfoRequest) (*pb.AccountInfoReply, error) {
+	var (
+		db   = ethdb.NewObjectDatabase(c.kv)
+		hash = common.BytesToHash(request.BlockHash)
+		addr = common.BytesToAddress(request.Account)
+	)
+
+	head, err := rawdb.ReadHeaderByHash(db, hash)
+	if err != nil {
+		return nil, err
+	}
+
+	acc, err := readAccountAtBlock(ctx, c.kv, addr, head.Number.Uint64())
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.AccountInfoReply{
+		Balance: acc.Balance.Bytes(),
+		Nonce:   i64tob(acc.Nonce),
+	}, nil
+}
+
+func (c *TxPoolControlServer) BlockStream(request *pb.BlockStreamRequest, stream pb.TxpoolControl_BlockStreamServer) error {
+	var (
+		db           = ethdb.NewObjectDatabase(c.kv)
+		lastHeader   = new(types.Header)
+		newRequestCh = make(chan *pb.BlockStreamRequest)
+		newHeadCh    = make(chan *types.Header)
+	)
+
+	queueRequest := func(ch chan *pb.BlockStreamRequest) error {
```

Comment:
You don't need all of that stuff with channels and go-routines here. The server is not supposed to read from the stream, it is supposed to only write into the stream. So, instead of doing all that, you can simply call `c.events.AddHeaderSubscription` and pass the function that will construct messages and send them to the stream using `stream.Send`. Unless I am missing something

Reply:
Actually, I probably wouldn't use `newHead` subscription for it either. I would embed all this into the TxPool stage. There, you have finer control of the reorgs, with `Unwind` phase where block are unwound and transactions need to be added to the pool, and `Forward` phase where transactions are taken away. Now sure if the protocol needs to be refactored for that though

Reply:
I will look into the TxPool stage and see how this can be embedded. As for your first comment, I believe the current interface is bidirectional so after a client subscribes to the provider they could send additional requests. Are you saying that those request should simply be dropped or do you think the interface should be changed to not support that?

Reply:
The current interface is bidirectional in the sense that there is also `txpool.proto` which we should probably merge into `txpool_control.proto` and invert all functions in `txpool.proto` to be functions returning streams or something. But in this case, the function `BlockStream` is not bidirectional. It takes one request as a parameter (not stream), and returns a stream of things. So one request generates many messages. In that sense, it is uni-directional. You may want to simplify the implementation, for example, by assuming and any further request replaces the previous one and closes all previous streams, or support many streams at the same time time (then, obviously, there will need to be some kind of fan-out logic to push the same message potentially to many streams). But no, you should not read from this stream as a server

Reply:
Also, if you implement the support for many streams at once, then you will need to detect when any stream is closed so that you can stop sending messages to it (but I believe you can simply see if `Send` returned an error and assume that the stream is gone or something - need to look into the docs for that)